### PR TITLE
Refactor code to allow for refinement

### LINKED ISSUE #173: Refactor CNS refinement modules to be executed after `topoaa`
With https://github.com/haddocking/haddock3/pull/152 we changed the data structure that `topoaa` exports which made it incompatible with the refinement modules.

These need to be refactored to be able to be executed after `topoaa`.

Additionally, the CNS code is expecting a single `input_pdb_filename_1`, meaning that the multiple molecules need to be merged as a single structure with multiple chains.

### PR DESCRIPTION
**Pull request by rvhonorato**
This PR refactors multiple modules to allow for refinement also adds it as an example.

The preferred way to submit things to refinement should be as an ensemble:

```python
# Ensemble to be refined
molecules = [
     "data/e2a-hpr_1GGR_ens.pdb",
    ]
```

If molecules are given separately, you need to use `merge`
```python
molecules =  [
    "data/e2a-hpr_1GGR_A.pdb",
    "data/e2a-hpr_1GGR_B.pdb"
    ]
# ...
[merge]
```

By using `merge` its possible to refine one ensemble against another, for example:
```python
molecules = [
    "receptor_ensemble.pdb",
    "smallmolecule_ensemble.pdb"
    ]
# ...
[merge]
```

`merge` will create all combinations (cartesian product), but we could also add a parameter to only do it pairwise.

I did touch the scoring modules yet, will do that in another PR since they are probably outdated.

* * *
Closes #173

### CODE REVIEW

File: src/haddock/gear/read_io.py
Comment (joaomcteixeira) [inline src/haddock/gear/read_io.py:1]: If this file is so heavily dependent on libs.libontology and is something we forsee a core part of the workflow, it should go to `libontology` directly.

File: src/haddock/libs/libpdb.py
Comment (joaomcteixeira) [inline src/haddock/libs/libpdb.py:61]: Honestly, I would use `pdbtools` here:

```python
from pdbtools import pdb_merge, pdb_tidy

def merge(pdb_file_list, new_pdb_file_path):
    input_fhs = (open(p) for p in pdb_file_list)
    _ = pdb_merge.run(input_fhs)
    merged = pdb_tidy.run(_)
    Path(new_pdb_file_path).write_text(os.linesep.join(merged))
```

I think there is no typo bug, the function should work as is. Also, your function is not closing the input files for you. While `pdb_merge` does it :+1:

File: src/haddock/gear/read_io.py
Comment (joaomcteixeira) [inline src/haddock/gear/read_io.py:8]: you don't need to create a list:

```python
if not all(isinstance(e, PDBFile) for e in io_output):
```

File: src/haddock/modules/refinement/emref/__init__.py
Comment (joaomcteixeira) [inline src/haddock/modules/refinement/emref/__init__.py:97]: Does this work? `check_balance` is not defined in the function definition.

File: src/haddock/modules/refinement/merge/__init__.py
Comment (joaomcteixeira) [inline src/haddock/modules/refinement/merge/__init__.py:46]: here use `enumerate(models_to_be_merged, start=1)` to avoid saving 0-indexed files to the system.

File: src/haddock/modules/refinement/merge/__init__.py
Comment (joaomcteixeira) [inline src/haddock/modules/refinement/merge/__init__.py:52]: Is `model_list` a list? Why not `topology = [e.topology for e in model_list]` . Maybe I am missing something.

File: src/haddock/modules/refinement/merge/__init__.py
Comment (joaomcteixeira) [inline src/haddock/modules/refinement/merge/__init__.py:54]: Is `topology` parameter receiving a list?

File: src/haddock/modules/sampling/rigidbody/__init__.py
Comment (joaomcteixeira) [inline src/haddock/modules/sampling/rigidbody/__init__.py:158]: This is redundant. `model.topology = topology_list` is enough.

File: src/haddock/modules/topology/topoaa/__init__.py
Comment (joaomcteixeira) [inline src/haddock/modules/topology/topoaa/__init__.py:100]: I see you didn't like my comments :-P

**Review by joaomcteixeira — CHANGES_REQUESTED** [review]
some changes for cleaning and improving

**Review by amjjbonvin — APPROVED** [review]
I will test once merged :-)

I see you are also including my CNS module changes into this PR.
Is this coming from my haddock3-example?

**Comment by rvhonorato** [comment]
> I will test once merged :-)
> 
> I see you are also including my CNS module changes into this PR. Is this coming from my haddock3-example?

yes! used it to test the refinement

**Comment by amjjbonvin** [comment]
So it will be part of your merge, right?

**Comment by rvhonorato** [comment]
@amjjbonvin caught a bug in which the capri metrics would be wrong, this was because `libpdb.get_pdb_file_suffix_variations` returned `list(abs_path.glob(f"{basename.stem}{sep}*{basename.suffix}"))` and [glob returns an arbitrary order list](https://stackoverflow.com/a/6773661), which resulted in the metric being evaluated over swapped chains. 

I added a sort 😉

File: src/haddock/gear/read_io.py
Comment (rvhonorato) [reply src/haddock/gear/read_io.py:1]: Your call

File: src/haddock/libs/libpdb.py
Comment (rvhonorato) [reply src/haddock/libs/libpdb.py:61]: this is `pdbtools`, check on the top `from pdbtools.pdb_merge import concatenate_files`

File: src/haddock/modules/refinement/emref/__init__.py
Comment (rvhonorato) [reply src/haddock/modules/refinement/emref/__init__.py:97]: ops forgot to delete it

File: src/haddock/modules/refinement/merge/__init__.py
Comment (rvhonorato) [reply src/haddock/modules/refinement/merge/__init__.py:52]: it can be a nested list if the model has multiple chains

File: src/haddock/modules/refinement/merge/__init__.py
Comment (rvhonorato) [reply src/haddock/modules/refinement/merge/__init__.py:54]: same as my comment above

File: src/haddock/modules/sampling/rigidbody/__init__.py
Comment (rvhonorato) [reply src/haddock/modules/sampling/rigidbody/__init__.py:158]: `topology_list` might be a nested list

**Comment by rvhonorato** [comment]
> So it will be part of your merge, right?

depends on who does it first! if we merge this pr you'll need to `update from master` but if you do it, then I need to update

**Comment by amjjbonvin** [comment]
One more thing about this refinement refactoring: In principle we want to refine a complex and generate multiple models (e.g. 20 or 50 per starting conformation), using a different random seed every time.

This means that when going from `topoaa` to e.g. `mdref` one must be able to specify the `sampling` parameter.

**Comment by amjjbonvin** [comment]
Also another strange behaviour which for some reason does not happen in the docking-protein-protein example is a duplication of PDB files when giving the models as separate molecules:

```
(haddock3) bash-3.2$ ls -1
total 9528
 176 e2a-hpr_1GGR_A.pdb
  72 e2a-hpr_1GGR_A_A.inp
5192 e2a-hpr_1GGR_A_A.out
 224 e2a-hpr_1GGR_A_A_haddock.pdb
 344 e2a-hpr_1GGR_A_A_haddock.psf
 224 e2a-hpr_1GGR_A_haddock.pdb
 104 e2a-hpr_1GGR_B.pdb
  72 e2a-hpr_1GGR_B_B.inp
2664 e2a-hpr_1GGR_B_B.out
 128 e2a-hpr_1GGR_B_B_haddock.pdb
 192 e2a-hpr_1GGR_B_B_haddock.psf
 128 e2a-hpr_1GGR_B_haddock.pdb
```

I am giving as input `e2a-hpr_1GGR_A.pdb` and `e2a-hpr_1GGR_B.pdb`, but as you can see four haddock pdb files are created.

**Comment by rvhonorato** [comment]
> This means that when going from `topoaa` to e.g. `mdref` one must be able to specify the `sampling` parameter.

Sure, but I could not find the name of this parameter in the CNS, which one should it be?

**Comment by amjjbonvin** [comment]
It is not a CNS parameter, but the python code must generate separate input files for each model request, even if only one starting model is provided.

I.e. it would be the sampling parameter of the other modules…

**Comment by rvhonorato** [comment]
> Also another strange behaviour which for some reason does not happen in the docking-protein-protein example is a duplication of PDB files when giving the models as separate molecules:

Ops, forgot to remove those as well, fixed in the last commit;

```bash
run1/00_topoaa [173●] » ls -1
e2a-hpr_1GGR_ens.pdb
e2a-hpr_1GGR_ens_1.pdb
e2a-hpr_1GGR_ens_1_A.inp
e2a-hpr_1GGR_ens_1_A.out
e2a-hpr_1GGR_ens_1_A_haddock.psf
e2a-hpr_1GGR_ens_1_B.inp
e2a-hpr_1GGR_ens_1_B.out
e2a-hpr_1GGR_ens_1_B_haddock.psf
e2a-hpr_1GGR_ens_1_haddock.pdb
e2a-hpr_1GGR_ens_2.pdb
e2a-hpr_1GGR_ens_2_A.inp
e2a-hpr_1GGR_ens_2_A.out
e2a-hpr_1GGR_ens_2_A_haddock.psf
e2a-hpr_1GGR_ens_2_B.inp
e2a-hpr_1GGR_ens_2_B.out
e2a-hpr_1GGR_ens_2_B_haddock.psf
e2a-hpr_1GGR_ens_2_haddock.pdb
io.json
```

PS: using Bash? 😄

**Comment by amjjbonvin** [comment]
Still - why split the files when they contain multiple chains?

And then merge then automatically, while if we provide separated chains you then need to add a merge step in the config file...

**Comment by rvhonorato** [comment]
> It is not a CNS parameter, but the python code must generate separate input files for each model request, even if only one starting model is provided. I.e. it would be the sampling parameter of the other modules…

Oh allright, we should name it in a way that it's clear that it is not a substitute for `seletop`, shall we do it per-model? then we can name it `conformation_sampling` ?

**Comment by rvhonorato** [comment]
Or expect people to properly split their multi-chained ensembles before inputting for scoring. 

I think that the way it is is more convenient since it supports both, but up to rule # 1!

File: src/haddock/libs/libpdb.py
Comment (joaomcteixeira) [reply src/haddock/libs/libpdb.py:61]: Yes, but your code repeats the `concatenate_files`. You are sending one file at the time, and getting one line at the time. Also `pdbmerge` is quite simple, if you don't want to use it, it fine. But the present code is using and desusing it :wink:

**Comment by joaomcteixeira** [comment]
Wait, don't merge yet. Rodrigo `sort` is nice, but we have already another function to help the `sort` further for `_1` like files. to avoid `_1`, `_10`, etc, and have `_1`, `_2`... We can add this here also to avoid bugs. Because I implemented that in another PR as well.

**Comment by amjjbonvin** [comment]
Sounds like a good choice.

But where would you input it? In the `mdref` module params?

> then we can name it conformation_sampling ?

**Comment by amjjbonvin** [comment]
But do you also split the files for scoring???

And then merge them again…

Why not simply give the complex directly to the module?  Why this complexity?

Again the CNS `topoaa` modules can handle models with multiple chains in them.

**Comment by rvhonorato** [comment]
Ok, my implementation relied heavily on the idea of supporting both multi-chain ensembles and single-chain ensembles and of course single-chain molecules and I changed the python engine around this.

So I am going to delete code this and do it again and taking into account your comments, thanks!

**Comment by rvhonorato** [comment]
> Why not simply give the complex directly to the module? Why this complexity?

To generically handle all possible input methods without cross-module dependency and support all input types: Single-chain Model, Single-chain ensemble, Multi-chain Model and Multi-chain ensemble.

But as I said, I already scratched this code and will try to think of a different implementation.

**Comment by joaomcteixeira** [comment]
Try to find a common path for all of them. For example, converting all to an ensemble and then moving from there. Or if not; these are just 4 cases. Implement 4 functions for each case. It is a well-defined problem. It won't grow beyond this.

**Pull request closed without merging**
